sklearn/datasets/california_housing.py

```diff
@@ -89,18 +90,19 @@ def fetch_california_housing(data_home=None, download_if_missing=True):
         makedirs(data_home)
     filepath = _pkl_filepath(data_home, TARGET_FILENAME)
     if not exists(filepath):
-        print('downloading Cal. housing from %s to %s' % (DATA_URL, data_home))
-        fhandle = urlopen(DATA_URL)
-        buf = BytesIO(fhandle.read())
-        zip_file = ZipFile(buf)
-        try:
-            cadata_fd = zip_file.open('cadata.txt', 'r')
-            cadata = BytesIO(cadata_fd.read())
-            # skip the first 27 lines (documentation)
-            cal_housing = np.loadtxt(cadata, skiprows=27)
-            joblib.dump(cal_housing, filepath, compress=6)
-        finally:
-            zip_file.close()
+        archive_path = os.path.join(data_home, ARCHIVE_NAME)
+        print('downloading Cal. housing from %s to %s' % (DATA_URL, archive_path))
+        urllib.urlretrieve(DATA_URL, archive_path)
+        fileobj = tarfile.open(archive_path, "r:gz").extractfile(
+            'CaliforniaHousing/cal_housing.data')
```

Comment:
Would it be possible to use a `urlopen` instead of `urlretrieve` to read the content of `DATA_URL` in memory and pass it as a pseudo-fileobj to `tarfile.open`?

Alternatively wrap it as was done in master: `BytesIO(urlopen(DATA_URL).read())`. This should make it possible to avoid creating a temporary file for the archive.